pkg/restic/backup.go

```diff
@@ -72,15 +59,197 @@ func (w *ResticWrapper) RunBackup(backupOption BackupOptions) (*BackupOutput, er
 	}
 
 	// Extract information from output of "stats" command
-	err = backupOutput.extractStatsInfo(out)
+	repoSize, err := extractStatsInfo(out)
 	if err != nil {
 		return nil, err
 	}
+	backupOutput.RepositoryStats.Size = repoSize
 
 	// Backup complete. Read current time and calculate total backup duration.
 	endTime := time.Now()
-	backupOutput.HostBackupStats.Duration = endTime.Sub(startTime).String()
-	backupOutput.HostBackupStats.Phase = api_v1beta1.HostBackupSucceeded
+	for idx := range backupOutput.HostBackupStats {
+		if backupOutput.HostBackupStats[idx].Hostname == backupOption.Host {
+			backupOutput.HostBackupStats[idx].Duration = endTime.Sub(startTime).String()
+			backupOutput.HostBackupStats[idx].Phase = api_v1beta1.HostBackupSucceeded
+		}
+	}
 
 	return backupOutput, nil
 }
+
+// RunParallelBackup runs multiple backup in parallel.
+// Host must be different for each backup.
+func (w *ResticWrapper) RunParallelBackup(backupOptions []BackupOptions, maxConcurrency int) (*BackupOutput, error) {
+
+	// Initialize restic repository if it does not exist
+	_, err := w.initRepositoryIfAbsent()
+	if err != nil {
+		return nil, err
+	}
+
+	// WaitGroup to wait until all go routine finish
+	wg := sync.WaitGroup{}
+	// concurrencyLimiter channel is used to limit maximum number simultaneous go routine
+	concurrencyLimiter := make(chan bool, maxConcurrency)
+	defer close(concurrencyLimiter)
+
+	var backupErr error
+	mu := sync.Mutex{} // use lock to avoid racing condition
+	backupOutput := &BackupOutput{}
+
+	for _, opt := range backupOptions {
+		// try to send message in concurrencyLimiter channel.
+		// if maximum allowed concurrent backup is already running, program control will stuck here.
+		concurrencyLimiter <- true
+
+		// starting new go routine. add it to WaitGroup
+		wg.Add(1)
+
+		// Start clock to measure session duration
+		startTime := time.Now()
+
+		go func(opt BackupOptions, start time.Time) {
+
+			// when this go routine completes it task, release a slot from the concurrencyLimiter channel
+			// so that another go routine can start. Also, tell the WaitGroup that it is done with its task.
+			defer func() {
+				<-concurrencyLimiter
+				wg.Done()
+			}()
+
+			// sh field in ResticWrapper is a pointer. we must not use same w in multiple go routine.
+			// otherwise they might enter in a racing condition.
+			nw := w.DeepCopy()
+
+			hostStats, err := nw.runBackup(opt)
+			if err != nil {
+				// acquire lock to make sure no other go routine is writing to backupErr variable
+				mu.Lock()
+				backupErr = errors.NewAggregate([]error{backupErr, err})
+				mu.Unlock()
+				return
+			}
+			hostStats.Duration = time.Now().Sub(start).String()
```

Comment:
Should we count time when the go routine was created or when it actually started running?